### PR TITLE
Include upgrade details in diagnostics bundle during ongoing upgrade

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR enhances the diagnostics bundle to include upgrade details when Agent is being upgraded.

## Why is it important?

To gain more visibility into the Agent upgrade process.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] ~I have added an integration test or an E2E test~

## How to test this PR locally

1. Build `elastic-agent` from this PR's branch and install it.
2. Run `elastic-agent diagnostics`, expand the resulting archive, and check the bundled `state.yaml` file. Verify that NO upgrade details are included in it by checking that there is no `upgrade_details` field.
3. In a new window, upgrade Elastic Agent to a version that hasn't been released yet.  This will ensure that the upgrade starts but remains stalled.
   ```
   sudo elastic-agent upgrade 8.13.0
   ```
4. In the original window, run `elastic-agent diagnostics` again, expand the resulting archive, and check the bundled `state.yaml` file. Verify that upgrade details are included in it by checking that there is an `upgrade_details` field.

## Related issues

- Relates to https://github.com/elastic/elastic-agent/issues/3119#issuecomment-1743638526

### CODE REVIEW

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 1]
```diff
@@ -799,12 +800,13 @@ func (c *Coordinator) DiagnosticHooks() diagnostics.Hooks {
...
-					FleetMessage: s.FleetMessage,
-					LogLevel:     s.LogLevel,
-					Components:   compStates,
+					State:          s.State,
+					Message:        s.Message,
+					FleetState:     s.FleetState,
+					FleetMessage:   s.FleetMessage,
+					LogLevel:       s.LogLevel,
+					Components:     compStates,
+					UpgradeDetails: s.UpgradeDetails,
```
Comment (pchila) [inline internal/pkg/agent/application/coordinator/coordinator.go:809]: When is this field present? Do we get details about the upgrade after a failure ? Or those get cleared as soon as the grace period is over ?

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 2]
```diff
@@ -781,12 +781,13 @@ func (c *Coordinator) DiagnosticHooks() diagnostics.Hooks {
...
-					FleetMessage string                 `yaml:"fleet_message"`
-					LogLevel     logp.Level             `yaml:"log_level"`
-					Components   []StateComponentOutput `yaml:"components"`
+					State          agentclient.State      `yaml:"state"`
+					Message        string                 `yaml:"message"`
+					FleetState     agentclient.State      `yaml:"fleet_state"`
+					FleetMessage   string                 `yaml:"fleet_message"`
+					LogLevel       logp.Level             `yaml:"log_level"`
+					Components     []StateComponentOutput `yaml:"components"`
+					UpgradeDetails *details.Details       `yaml:"upgrade_details,omitempty"`
```
Comment (pchila) [inline internal/pkg/agent/application/coordinator/coordinator.go:790]: I don't see any changes in the tests related to this change... could you please add a unit test to document what we can expect in the hook output when UpgradeDetails is populated ?

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 1 above)
Comment (ycombinator) [reply internal/pkg/agent/application/coordinator/coordinator.go:809]: > When is this field present? 

This field is populated when the upgrade process starts: https://github.com/elastic/elastic-agent/blob/2a93086d5c7695f9a08e118667dac0da677157c6/internal/pkg/agent/application/coordinator/coordinator.go#L475-L476

> Do we get details about the upgrade after a failure ?

Yes, if the upgrade process fails at any point, this field will continue to remain populated. 

> Or those get cleared as soon as the grace period is over ?

I assume you are referring to the Upgrade Watcher's grace period.

Exactly when we clear it out after that is not yet implemented. We need to decide whether we want to clear it out based on some timeout (maybe the grace period or maybe something even beyond that) or due to some user intervention (which could be retrying the upgrade or something else just to clear the upgrade details, not sure yet) or either (whichever happens first).

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 2 above)
Comment (ycombinator) [reply internal/pkg/agent/application/coordinator/coordinator.go:790]: Thanks.  Added in 02625e831564e5371bdaf6ca81771216f90df926.

Comment (ycombinator) [reply internal/pkg/agent/application/coordinator/coordinator.go:800]: Thanks.  Added in 02625e831564e5371bdaf6ca81771216f90df926.

**Review by blakerouse — APPROVED** [review]
Nice! Great this is added to the diagnostics.

**Review by pchila — APPROVED** [review]
LGTM

**Pull request merged**